deprecate inlineCallbacks (#1374)
twisted/twisted: docs/core/howto/defer-intro.rst

Perhaps we want something that looks a little like this::

    async with pod_bay_doors.open():

>         await pod.launch()

Reviewer: Maybe the example is easier to understand without the `with` statement:
```py
        await pod_bay_doors.open()
        pod.launch()
```
The `with` block might be better if this were actual code, since it could make sure the doors stay open until the pod has launched and close afterwards. But we're just illustrating a concept, not writing code for a spaceship ;)

Author: I think it's important to show all the features up front as early as possible in the document. But I'm happy to rollback on this change to get it landed

Reviewer: It's not a big issue for me; if no-one else has a clear preference you can go with your solution.

Thank goodness we can!

> Using coroutines
> ~~~~~~~~~~~~~~~~

Author: ```suggestion
Coroutines with async/await 
~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

> await" functionality can be used with Deferreds by the use of :api:`twisted.internet.defer.ensureDeferred <ensureDeferred>`.

Author: oops I trimmed a bit much there

```
On Python 3.5 and higher, the :pep:`492` ("Coroutines with async and await syntax") "await" functionality can be used with Deferreds by the use of :api:`twisted.internet.defer.ensureDeferred <ensureDeferred>`.
```

```suggestion
"async"/ "await" coroutine functionality can be used with Deferreds by the use of :api:`twisted.internet.defer.ensureDeferred <ensureDeferred>`.
```

Our exception handling is simplified because we can use Python's familiar ``try`` / ``except`` syntax for handling ``ConnectionError``\ s.

When writing coroutines, you do not need to use :api:`twisted.internet.defer.ensureDeferred <ensureDeferred>` when you are writing a coroutine which calls other coroutines which await on Deferreds; you can just ``await`` on it directly.
For example:

.. code-block:: python3

    async def foo():
        return await someFunctionThatReturnsADeferred()

    async def bar():
        return await someOtherDeferredFunction() + await foo()

    def myDeferredReturningFunction():
        return ensureDeferred(coro())

Even though Deferreds were used in both coroutines, only ``bar`` had to be wrapped in :api:`twisted.internet.defer.ensureDeferred <ensureDeferred>` to return a Deferred.

A low level solution: Deferred
------------------------------

At this point you know everything you need to know about consuming asynchronous APIs. The ``ensureDeferred`` is built on a low level api described below. It's generally recommended to exlusively use coroutines, but sometimes you'll see code using the lower-level raw-deferred interface.

> Twisted tackles this problem with :api:`twisted.internet.defer.Deferred <Deferred>`\s, a type of object designed to do one thing, and one thing only: encode an order of execution separately from the order of lines in Python source code.

Author: ```suggestion
Twisted tackles this problem with :api:`twisted.internet.defer.Deferred <Deferred>`\s, a type of object designed to do one thing, and one thing only: to bridge low-level callback-based code with high-level async/await code.
```

Otherwise, ``g`` will be passed the asynchronous equivalent of the return value of ``f()`` (i.e. ``y``).

Coroutines with async/await
~~~~~~~~~~~~~~~~~~~~~~~~~~~

> 

Reviewer: There are also some changes that could be made to the coroutines section:

- Twisted requires Python 3.5+ now, so mentioning that version requirement for coroutines is superfluous
- the description states "It is similar to `inlineCallbacks`", but the referenced text has been deleted, so it's probably a good idea to migrate relevant parts of it to this section

Otherwise, ``g`` will be passed the asynchronous equivalent of the return value of ``f()`` (i.e. ``y``).

> Inline callbacks - using 'yield'

Reviewer 2: You cannot remove this section.   In a PR which deprecates a feature, you cannot delete things (even documentation).  The deprecation and removal policy is described here: https://twistedmatrix.com/documents/current/core/development/policy/compatibility-policy.html

`inlineCallbacks()` have been around for a while, and there are actually a lot of projects and companies which still use them.  Some of these projects/companies may take a while (years) to migrate from `inlineCallbacks()` to the new `await` keyword.

Author: Can I put it somewhere else? I don't think it should be in the introduction

Reviewer 2: No
